OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs

```diff
@@ -75,11 +89,52 @@ public void ResolveOrder(Actor self, Order order)
 				return;
 
 			if (isUpgraded)
-				foreach (var up in info.Upgrades)
-					manager.RevokeUpgrade(self, up, this);
+			{
+				// Play undeploy animation and after that revoke the upgrades
+				self.QueueActivity(false, new CallFunc(() =>
+				{
+					if (!string.IsNullOrEmpty(info.UndeploySound))
+						Sound.Play(info.UndeploySound, self.CenterPosition);
+
+					if (string.IsNullOrEmpty(info.DeployAnimation))
+					{
+						RevokeUpgrades();
+						return;
+					}
+
+					var render = self.TraitOrDefault<ISpriteBody>();
```

Comment:
Nobody said this is the final version of the trait, but without a test case I can't reliably make it support every possible scenario. Also it would be much easier to digest if it's not all at once, I assume.